server/services/v1/auth.go

```diff
@@ -116,3 +283,91 @@ func (h *authService) RegisterGRPC(grpc *grpc.Server) error {
 	api.RegisterAuthServer(grpc, h)
 	return nil
 }
+
+func getCurrentSub(ctx context.Context) (string, error) {
+	// further filter for this particular user
+	token, err := request.GetAccessToken(ctx)
+	if err != nil {
+		return "", api.Errorf(api.Code_INTERNAL, "Failed to list applications: reason = %s", err.Error())
+	}
+	return token.Sub, nil
+}
+
+func getAccessTokenUsingRefreshToken(req *api.GetAccessTokenRequest, a *Auth0) (*api.GetAccessTokenResponse, error) {
+	data := url.Values{
+		"refresh_token": {req.RefreshToken},
+		"client_id":     {a.authConfig.ClientId},
+		"grant_type":    {refreshToken},
+		"scope":         {scope},
+	}
+	resp, err := http.PostForm(a.authConfig.ExternalTokenURL, data)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to get access token: reason = %s", err.Error())
+	}
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to get access token: reason = %s", err.Error())
+	}
+	bodyStr := string(body)
+	if resp.StatusCode == 200 {
+		getAccessTokenResponse := api.GetAccessTokenResponse{}
+
+		err = json.Unmarshal([]byte(bodyStr), &getAccessTokenResponse)
+		if err != nil {
+			return nil, api.Errorf(api.Code_INTERNAL, "Failed to parse external response: reason = %s", err.Error())
+		}
+		return &getAccessTokenResponse, nil
+	}
+	log.Error().Msgf("Auth0 response status code=%d", resp.StatusCode)
+	return nil, api.Errorf(api.Code_INTERNAL, "Failed to get access token: reason = %s", bodyStr)
```

Comment:
I think we need to make sure we are returning the correct error code that we have received in the response otherwise everything will be an internal error. But you can fix it iteratively.

Reply:
updated the error handling to convey the closest status code back.